src/sensr_client.cpp

```diff
@@ -1,119 +1,134 @@
 #include "sensr_client.h"
+
+#include <algorithm>
+
+#include "sensr_message_listener.h"
 #include "websocket/websocket_endpoint.h"
 #include "websocket/websocket_secure_endpoint.h"
-#include "sensr_message_listener.h"
-#include <algorithm>
 
-namespace sensr
-{
-  Client::Client(const std::string& address, const std::string& cert_path) 
+namespace sensr {
+  Client::Client(const std::string& address, const std::string& cert_path)
       : address_(address), use_ssl_(!cert_path.empty()), is_reconnecting_(false) {
     if (use_ssl_) {
       output_endpoint_.reset(new WebSocketSecureEndPoint(cert_path));
       point_endpoint_.reset(new WebSocketSecureEndPoint(cert_path));
     } else {
       output_endpoint_.reset(new WebSocketEndPoint());
       point_endpoint_.reset(new WebSocketEndPoint());
-    }    
+    }
   }
 
-  Client::~Client()
-  {
+  Client::~Client() {
     for (const auto& listener : listeners_) {
       UnsubscribeMessageListener(listener);
     }
-    StopReconnecting();
+
+    io_context_.stop();
+    if (reconnection_thread_.joinable()) {
+      reconnection_thread_.join();
+    }
   }
 
   void Client::Reconnect() {
     if (!is_reconnecting_) {
-      StopReconnecting();
-      reconnection_thread_ = std::thread([this]() {
-        is_reconnecting_ = true;
-        auto temp = listeners_;
-        size_t reconnection_attempts_ = 0;
-        while (reconnection_attempts_++ < Client::kMaxReconnectTrialCount) {
-          INFO_LOG("Reconnecting...");
-          for (const auto& listener : listeners_) {
-            UnsubscribeMessageListener(listener);
-          }
-          for (const auto& listener : temp) {
-            SubscribeMessageListener(listener);
-          }
-          bool is_output_connected = true;
-          bool is_point_connected = true;
-          if (IsResultListening()) {
-            if (!output_endpoint_->IsConnected()) {
-              is_output_connected = false;
-            }
-          }
-          if (IsPointListening()) {
-            if (!point_endpoint_->IsConnected()) {
-              is_point_connected = false;
-            }
-          }
-
-          if (is_output_connected && is_point_connected) {
-            INFO_LOG("Reconnected!");
-            break;
-          } else if (!is_reconnecting_) {
-            INFO_LOG("Stop reconnection!");
-            break;
-          }
-          std::this_thread::sleep_for(std::chrono::seconds(1));
-        }
-        is_reconnecting_ = false;
+      if (reconnection_thread_.joinable()) {
+        reconnection_thread_.join();
+      }
+
+      is_reconnecting_ = true;
+      asio::post(io_context_, [this]() { reconnection_async(); });
+      reconnection_thread_ = std::thread([this] {
+        io_context_.restart();
+        io_context_.run();
```

Comment:
Doesn't `restart` flush the async reconnection?
```suggestion
        io_context_.run();
```

Reply:
I am not sure `restart` flushing left task in the queue or not. When I checked the function in the code, restart() just change state of `stopped_` variable in context's scheduler to false. Once io_context's run() function returned, the state of `stopped_` changed to true. And it literally stops. For using it multiple times, we need to restart it. Just in case, I will move restart function before posting the job to context.